File: src/main/java/ch/obermuhlner/astro/gradient/filter/CopyFilter.java
package ch.obermuhlner.astro.gradient.filter;

import ch.obermuhlner.astro.image.color.ColorModel;
import ch.obermuhlner.astro.image.DoubleImage;

public class CopyFilter implements Filter {

  private final ColorModel model;

  public CopyFilter(ColorModel model) {
    this.model = model;
  }

  @Override
  public DoubleImage filter(DoubleImage source, DoubleImage target, int width, int height) {
    double[] samples = new double[3];
    for (int dy = 0; dy < height; dy++) {
      for (int dx = 0; dx < width; dx++) {
        if (source.isInside(dx, dy)) {
          source.getPixel(dx, dy, model, samples);
        } else {
          samples[0] = 0;
          samples[1] = 0;
          samples[2] = 0;
        }

        target.setPixel(dx, dy, model, samples);
      }
    }
    return target;
  }

  @Override
  public String toString() {
    return "CopyFilter";
  }
}

File: src/main/java/ch/obermuhlner/astro/gradient/filter/CopyFilter.kt
package ch.obermuhlner.astro.gradient.filter

import ch.obermuhlner.astro.image.DoubleImage
import ch.obermuhlner.astro.image.color.ColorModel

class CopyFilter constructor(private val model: ColorModel) : Filter {
    override fun filter(source: DoubleImage, target: DoubleImage, width: Int, height: Int): DoubleImage {
        val samples = DoubleArray(3)
        for (dy in 0 until height) {
            for (dx in 0 until width) {
                if (source.isInside(dx, dy)) {
                    source.getPixel(dx, dy, model, samples)
                } else {
                    samples[0] = 0.0
                    samples[1] = 0.0
                    samples[2] = 0.0
                }
                target.setPixel(dx, dy, model, samples)
            }
        }
        return target
    }

    override fun toString(): String {
        return "CopyFilter"
    }
}